// @jacobmischka/ical-merger/dist/components/ActiveEvent.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

import _JSXStyle from "styled-jsx/style";

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import React, { Fragment, createRef } from 'react';
import PropTypes from 'prop-types';
import Color from 'color';
import Calendar from 'react-feather/dist/icons/calendar.js';
import Clock from 'react-feather/dist/icons/clock.js';
import MapPin from 'react-feather/dist/icons/map-pin.js';
import User from 'react-feather/dist/icons/user.js';
import Users from 'react-feather/dist/icons/users.js';
import CalendarEvent from './CalendarEvent.js';
import AddToCalendar from './AddToCalendar.js';
import EventLocation from './EventLocation.js';
import { BREAKPOINTS, OPACITIES, COLORS } from '../constants.js';
import { rgbaOverRgb } from '../utils.js';

var ActiveEvent = /*#__PURE__*/function (_CalendarEvent) {
  _inherits(ActiveEvent, _CalendarEvent);

  var _super = _createSuper(ActiveEvent);

  function ActiveEvent(props) {
    var _this;

    _classCallCheck(this, ActiveEvent);

    _this = _super.call(this, props);
    _this.state = {
      expanded: false
    };
    _this.getEventDate = _this.getEventDate.bind(_assertThisInitialized(_this));
    _this.handleOutsideClick = _this.handleOutsideClick.bind(_assertThisInitialized(_this));
    _this.handleClick = _this.handleClick.bind(_assertThisInitialized(_this));
    _this.addLinkTargets = _this.addLinkTargets.bind(_assertThisInitialized(_this));
    _this.descriptionRef = /*#__PURE__*/createRef();
    return _this;
  }

  _createClass(ActiveEvent, [{
    key: "getEventDate",
    value: function getEventDate() {
      var event = this.props.event;

      if (!event.end) {
        event.end = event.start;
      }

      var sameDay = event.start.isSame(event.end, 'day');
      var sameDayAllDay = event.allDay && event.start.isSame(event.end.clone().subtract(1, 'day'));

      if (sameDay || sameDayAllDay) {
        return event.start.format('ll');
      } else {
        var startDate, endDate;

        if (event.start.isSame(event.end, 'year')) {
          startDate = event.start.format('MMM D');
          endDate = event.end.format('ll');
        } else {
          var dateFormat = 'll';
          startDate = event.start.format(dateFormat);
          endDate = event.end.format(dateFormat);
        }

        return /*#__PURE__*/React.createElement("span", null, /*#__PURE__*/React.createElement("time", {
          className: "start-date",
          dateTime: event.start.toISOString()
        }, startDate), /*#__PURE__*/React.createElement("span", null, " \u2013 "), /*#__PURE__*/React.createElement("time", {
          className: "end-date",
          dateTime: event.end.toISOString()
        }, endDate));
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          event = _this$props.event,
          showFooter = _this$props.showFooter;
      var style;

      if (!this.state.expanded) {
        var rect = this.props.originalElement ? this.props.originalElement.getBoundingClientRect() : {
          top: 1,
          right: 1,
          bottom: 1,
          left: 1,
          width: 0,
          height: 0
        };
        var left = rect.left - 1;
        var top = rect.top - 1;
        var backgroundColor = rgbaOverRgb(Color(event.color).alpha(0.3).array());
        var borderColor = event.color;
        style = {
          width: rect.width * 2,
          height: rect.height * 2,
          transform: "translate(calc(".concat(left, "px - 25%), calc(").concat(top, "px - 25%)) scale(0.5)"),
          backgroundColor: backgroundColor,
          border: "2px solid ".concat(borderColor)
        };
      }

      var eventTime = this.getEventTime();
      var eventDate = this.getEventDate();
      var className = this.getClassName();
      className += ' active-event';
      if (this.state.expanded) className += ' expanded';
      var headerStyle;

      if (this.state.expanded) {
        headerStyle = {
          borderBottom: "5px solid ".concat(event.color)
        };
      }

      var hasPresenters = event.presenters && event.presenters.length > 0;
      return /*#__PURE__*/React.createElement("div", {
        key: "active-event",
        style: style,
        ref: function ref(div) {
          return _this2.container = div;
        },
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + (className || "")
      }, /*#__PURE__*/React.createElement("header", {
        style: headerStyle,
        ref: function ref(header) {
          return _this2.header = header;
        },
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]])
      }, /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-date-time"
      }, /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-date"
      }, /*#__PURE__*/React.createElement(Calendar, null), eventDate), /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-time"
      }, /*#__PURE__*/React.createElement(Clock, null), eventTime)), /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-title"
      }, /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-calendar"
      }, event.calendar.calname), event.title), /*#__PURE__*/React.createElement("button", {
        type: "button",
        onClick: this.handleClick,
        title: "Close active event",
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "close"
      }, "\xD7")), (event.location || hasPresenters) && /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta"
      }, event.location && /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-pair"
      }, /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-label"
      }, /*#__PURE__*/React.createElement(MapPin, null), "Location"), /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-value event-location"
      }, /*#__PURE__*/React.createElement(EventLocation, {
        location: event.location
      }))), hasPresenters && /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-pair"
      }, /*#__PURE__*/React.createElement("span", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-label"
      }, event.presenters.length === 1 ? /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement(User, null), "Presenter") : /*#__PURE__*/React.createElement(Fragment, null, /*#__PURE__*/React.createElement(Users, null), "Presenters")), /*#__PURE__*/React.createElement("ul", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-meta-value event-presenters"
      }, event.presenters.map(function (presenter) {
        return /*#__PURE__*/React.createElement("li", {
          key: presenter.name,
          className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]])
        }, presenter.name);
      })))), event.description && /*#__PURE__*/React.createElement("div", {
        ref: this.descriptionRef,
        dangerouslySetInnerHTML: this.markupDescription(event.description),
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]]) + " " + "event-desc"
      }), showFooter && /*#__PURE__*/React.createElement("footer", {
        className: _JSXStyle.dynamic([["4112202273", [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]]])
      }, /*#__PURE__*/React.createElement(AddToCalendar, {
        event: event
      })), /*#__PURE__*/React.createElement(_JSXStyle, {
        id: "4112202273",
        dynamic: [OPACITIES.TEXT.PRIMARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.DISABLED, OPACITIES.TEXT.SECONDARY, COLORS.SHADOW, COLORS.SHADOW, COLORS.BACKGROUND, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.SECONDARY, OPACITIES.TEXT.PRIMARY, BREAKPOINTS.SMALL_DESKTOP]
      }, [".active-event.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;font-family:'Noto Sans',sans-serif;color:rgba(0,0,0,".concat(OPACITIES.TEXT.PRIMARY, ");padding:0.5em;margin:1px;cursor:pointer;box-sizing:border-box;position:fixed;left:0;top:0;width:90vw;height:90vh;font-size:1.5em;overflow:hidden;z-index:100;cursor:auto;visibility:hidden;}"), "header.__jsx-style-dynamic-selector,footer.__jsx-style-dynamic-selector{-webkit-box-flex:0;-webkit-flex-grow:0;-ms-flex-positive:0;flex-grow:0;background-color:transparent;}", ".event-calendar.__jsx-style-dynamic-selector{display:none;font-size:0.7em;color:rgba(0,0,0,".concat(OPACITIES.TEXT.SECONDARY, ");}"), ".event-date-time.__jsx-style-dynamic-selector{margin:0;text-align:center;}", ".event-date.__jsx-style-dynamic-selector{display:none;text-transform:none;}", ".start-date.__jsx-style-dynamic-selector,.end-date.__jsx-style-dynamic-selector{white-space:nowrap;}", ".event-time.__jsx-style-dynamic-selector{margin:0;text-transform:uppercase;}", ".event-title.__jsx-style-dynamic-selector{word-wrap:break-word;max-width:80%;}", ".event-meta.__jsx-style-dynamic-selector{display:none;}", ".event-location.__jsx-style-dynamic-selector{text-align:left;color:rgba(0,0,0,".concat(OPACITIES.TEXT.SECONDARY, ");}"), ".event-desc.__jsx-style-dynamic-selector{-webkit-box-flex:1;-webkit-flex-grow:1;-ms-flex-positive:1;flex-grow:1;-webkit-flex-shrink:1;-ms-flex-negative:1;flex-shrink:1;margin:0;font-size:0.75em;white-space:pre-wrap;word-break:break-word;}", ".event.all-day.__jsx-style-dynamic-selector .event-time.__jsx-style-dynamic-selector{text-align:center;background:#bfbfbf;border-radius:2px;padding:0.15em 0.5em;}", "header.__jsx-style-dynamic-selector .close.__jsx-style-dynamic-selector{position:absolute;right:0;top:0;cursor:pointer;font-size:1em;background:none;border:none;outline:none;color:rgba(0,0,0,".concat(OPACITIES.TEXT.DISABLED, ");padding:0.1em;margin:0.1em;line-height:0.65;}"), "header.__jsx-style-dynamic-selector .close.__jsx-style-dynamic-selector:hover{color:rgba(0,0,0,".concat(OPACITIES.TEXT.SECONDARY, ");}"), ".active-event.expanded.__jsx-style-dynamic-selector{font-size:1.5em;background-color:#fdfdfd;border:1px solid grey;padding:0;-webkit-transform:translate(50vw,50vh) translate(-50%,-50%);-ms-transform:translate(50vw,50vh) translate(-50%,-50%);transform:translate(50vw,50vh) translate(-50%,-50%);box-shadow:0 0 20px 0 ".concat(COLORS.SHADOW, ";visibility:visible;}"), ".expanded.__jsx-style-dynamic-selector header.__jsx-style-dynamic-selector,.expanded.__jsx-style-dynamic-selector footer.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-flex-wrap:wrap;-ms-flex-wrap:wrap;flex-wrap:wrap;box-shadow:0 0 5px 0 ".concat(COLORS.SHADOW, ";background-color:").concat(COLORS.BACKGROUND, ";overflow:auto;}"), ".expanded.__jsx-style-dynamic-selector header.__jsx-style-dynamic-selector{-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;padding:0.5em;}", ".expanded.__jsx-style-dynamic-selector footer.__jsx-style-dynamic-selector{-webkit-box-pack:space-around;-webkit-justify-content:space-around;-ms-flex-pack:space-around;justify-content:space-around;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;text-align:center;}", ".expanded.__jsx-style-dynamic-selector header.__jsx-style-dynamic-selector svg,.expanded.__jsx-style-dynamic-selector .event-meta.__jsx-style-dynamic-selector svg{color:rgba(0,0,0,".concat(OPACITIES.TEXT.SECONDARY, ");height:1em;width:1em;vertical-align:baseline;margin-right:0.25em;overflow:visible;}"), ".expanded.__jsx-style-dynamic-selector .event-title.__jsx-style-dynamic-selector,.expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector{margin:0.25em;}", ".expanded.__jsx-style-dynamic-selector .event-title.__jsx-style-dynamic-selector{-webkit-order:1;-ms-flex-order:1;order:1;-webkit-box-flex:1;-webkit-flex-grow:1;-ms-flex-positive:1;flex-grow:1;max-width:100%;}", ".expanded.__jsx-style-dynamic-selector .event-calendar.__jsx-style-dynamic-selector{display:block;}", ".expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector{-webkit-order:2;-ms-flex-order:2;order:2;font-size:0.9em;}", ".expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector svg{vertical-align:middle;}", ".expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector svg{margin-right:0.5em;}", ".expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector .event-date.__jsx-style-dynamic-selector,.expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector .event-time.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;}", ".expanded.__jsx-style-dynamic-selector .event-date.__jsx-style-dynamic-selector,.expanded.__jsx-style-dynamic-selector .event-time.__jsx-style-dynamic-selector{display:block;margin:0.1em 0.25em;white-space:nowrap;}", ".expanded.__jsx-style-dynamic-selector .event-meta.__jsx-style-dynamic-selector{display:block;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-flex-wrap:wrap;-ms-flex-wrap:wrap;flex-wrap:wrap;-webkit-align-items:space-around;-webkit-box-align:space-around;-ms-flex-align:space-around;align-items:space-around;background-color:#fafafa;border-bottom:1px solid #ccc;font-size:0.75em;}", ".expanded.__jsx-style-dynamic-selector .event-meta-pair.__jsx-style-dynamic-selector{-webkit-flex:1 1;-ms-flex:1 1;flex:1 1;padding:0.75em;border:1px solid #ccc;}", ".expanded.__jsx-style-dynamic-selector .event-meta-pair.__jsx-style-dynamic-selector .event-meta-label.__jsx-style-dynamic-selector{color:rgba(0,0,0,".concat(OPACITIES.TEXT.SECONDARY, ");white-space:nowrap;}"), ".expanded.__jsx-style-dynamic-selector .event-meta-pair.__jsx-style-dynamic-selector .event-meta-value.__jsx-style-dynamic-selector{color:rgba(0,0,0,".concat(OPACITIES.TEXT.PRIMARY, ");}"), ".expanded.__jsx-style-dynamic-selector .event-desc.__jsx-style-dynamic-selector{padding:1em 2em 2em;overflow-y:auto;}", ".expanded.__jsx-style-dynamic-selector .event-location.__jsx-style-dynamic-selector{display:block;}", ".expanded.__jsx-style-dynamic-selector .event-presenters.__jsx-style-dynamic-selector{display:block;padding:0;margin:0;}", ".expanded.__jsx-style-dynamic-selector .event-presenters.__jsx-style-dynamic-selector li.__jsx-style-dynamic-selector{display:inline;}", ".expanded.__jsx-style-dynamic-selector .event-presenters.__jsx-style-dynamic-selector li.__jsx-style-dynamic-selector:not(:last-child)::after{content:', ';}", "@media (min-width:".concat(BREAKPOINTS.SMALL_DESKTOP, "px){.active-event.__jsx-style-dynamic-selector{-webkit-transition-duration:0.15s;transition-duration:0.15s;-webkit-transition-property:left,top,width,height,-webkit-transform,background-color,border;-webkit-transition-property:left,top,width,height,transform,background-color,border;transition-property:left,top,width,height,transform,background-color,border;max-width:90vw;max-height:90vh;width:auto;height:auto;visibility:visible;}header.__jsx-style-dynamic-selector,footer.__jsx-style-dynamic-selector{-webkit-transition-duration:0.15s;transition-duration:0.15s;-webkit-transition-property:background-color,border;transition-property:background-color,border;}.active-event.expanded.__jsx-style-dynamic-selector{font-size:2em;}.event-date-time.__jsx-style-dynamic-selector{text-align:right;}.expanded.__jsx-style-dynamic-selector header.__jsx-style-dynamic-selector{-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;-webkit-flex-wrap:nowrap;-ms-flex-wrap:nowrap;flex-wrap:nowrap;text-align:left;}.expanded.__jsx-style-dynamic-selector .event-title.__jsx-style-dynamic-selector{margin-right:1em;font-size:1.25em;}.expanded.__jsx-style-dynamic-selector .event-meta.__jsx-style-dynamic-selector{font-size:0.6em;}.expanded.__jsx-style-dynamic-selector .event-meta-pair.__jsx-style-dynamic-selector{padding:0.75em 1em;}.expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector{-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:end;-webkit-justify-content:flex-end;-ms-flex-pack:end;justify-content:flex-end;}.expanded.__jsx-style-dynamic-selector .event-date-time.__jsx-style-dynamic-selector{margin-left:1em;}}")]));
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this3 = this;

      window.requestAnimationFrame(function () {
        window.requestAnimationFrame(function () {
          _this3.setState({
            expanded: true
          });

          if ('parentIFrame' in window) {
            window.parentIFrame.getPageInfo(function (parentPage) {
              var middleOfParentViewport = parentPage.scrollTop + parentPage.clientHeight / 2 - parentPage.offsetTop;
              var y = middleOfParentViewport > parentPage.clientHeight / 2 ? middleOfParentViewport < window.innerHeight - parentPage.clientHeight / 2 ? "".concat(middleOfParentViewport, "px") : '100vh - 50% - 10px' : '50% + 10px';
              var containerMaxHeight = parentPage.clientHeight * 0.9 - 50;
              _this3.container.style.maxHeight = "".concat(containerMaxHeight, "px");
              _this3.container.style.transform = "translate(calc(50vw - 50%), calc(".concat(y, " - 50%))");
            });
          }

          document.addEventListener('click', _this3.handleOutsideClick);
        });
      });
      this.addLinkTargets();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.addLinkTargets();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      document.removeEventListener('click', this.handleOutsideClick);
    }
  }, {
    key: "handleClick",
    value: function handleClick() {
      var _this4 = this;

      window.requestAnimationFrame(function () {
        _this4.setState({
          expanded: false
        }, function () {
          if ('parentIFrame' in window) {
            window.parentIFrame.getPageInfo(false);
          }

          var timeout = window.innerWidth >= BREAKPOINTS.SMALL_DESKTOP ? 140 : 10;
          window.setTimeout(_this4.props.onClose, timeout);
        });
      });
    }
  }, {
    key: "handleOutsideClick",
    value: function handleOutsideClick(event) {
      var _this5 = this;

      if (event.defaultPrevented) return;
      window.requestAnimationFrame(function () {
        var rect = _this5.container.getBoundingClientRect();

        if (event.clientX < rect.left || event.clientX > rect.right || event.clientY < rect.top || event.clientY > rect.bottom) _this5.handleClick();
      });
    }
  }, {
    key: "addLinkTargets",
    value: function addLinkTargets() {
      var container = this.descriptionRef.current;

      if (container) {
        var anchors = container.querySelectorAll('a');

        var _iterator = _createForOfIteratorHelper(anchors),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var anchor = _step.value;

            if (!anchor.target) {
              anchor.target = '_blank';
            }

            if (!anchor.rel) {
              anchor.rel = 'noopener noreferrer';
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      }
    }
  }]);

  return ActiveEvent;
}(CalendarEvent);

export { ActiveEvent as default };
ActiveEvent.propTypes = {
  event: PropTypes.object.isRequired,
  originalElement: PropTypes.object,
  onClose: PropTypes.func,
  showFooter: PropTypes.bool
};
ActiveEvent.defaultProps = {
  showFooter: false
};
