// @jacobmischka/ical-merger/dist/components/CalendarLegend.js
import _JSXStyle from "styled-jsx/style";
import React from 'react';
import PropTypes from 'prop-types';
import Color from 'color';
import { COLORS } from '../constants.js';
var legendBorderColor = new Color(COLORS.GREY).alpha(0.45);

var CalendarLegend = function CalendarLegend(props) {
  if (!props.calendars || props.calendars.length < 1) return null;
  var calendarsInGroup = props.calendars.map(function (calendar) {
    return /*#__PURE__*/React.createElement("li", {
      key: "in-group-".concat(calendar.calname),
      className: "jsx-902230863" + " " + "legend-list-item"
    }, /*#__PURE__*/React.createElement("span", {
      style: {
        backgroundColor: Color(calendar.color).alpha(0.3).rgb().string(),
        border: "1px solid ".concat(calendar.color)
      },
      className: "jsx-902230863" + " " + "calendar-legend-color"
    }), calendar.calname, /*#__PURE__*/React.createElement(_JSXStyle, {
      id: "902230863"
    }, [".legend-list-item.jsx-902230863{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;}", ".calendar-legend-color.jsx-902230863{display:inline-block;width:1em;height:1em;margin:0 0.5em 0 0;}"]));
  });
  return /*#__PURE__*/React.createElement("div", {
    className: _JSXStyle.dynamic([["4283081990", [legendBorderColor]]]) + " " + "calendar-legend-container"
  }, /*#__PURE__*/React.createElement("div", {
    className: _JSXStyle.dynamic([["4283081990", [legendBorderColor]]]) + " " + "calendar-legend"
  }, /*#__PURE__*/React.createElement("span", {
    className: _JSXStyle.dynamic([["4283081990", [legendBorderColor]]]) + " " + "legend-title"
  }, "Calendars in ".concat(props.calname)), /*#__PURE__*/React.createElement("ul", {
    className: _JSXStyle.dynamic([["4283081990", [legendBorderColor]]]) + " " + "legend-list"
  }, calendarsInGroup)), /*#__PURE__*/React.createElement(_JSXStyle, {
    id: "4283081990",
    dynamic: [legendBorderColor]
  }, [".calendar-legend-container.__jsx-style-dynamic-selector{margin:2em;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;}", ".calendar-legend.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-box-pack:justify;-webkit-justify-content:space-between;-ms-flex-pack:justify;justify-content:space-between;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;background-color:#fafafa;padding:1em;border:1px solid ".concat(legendBorderColor, ";}"), ".legend-title.__jsx-style-dynamic-selector{text-align:center;font-size:1.1em;}", ".legend-list.__jsx-style-dynamic-selector{padding:0 1em;}", "@media print{.calendar-legend-container.__jsx-style-dynamic-selector{font-size:0.65em;}}"]));
};

export default CalendarLegend;
CalendarLegend.propTypes = {
  calendars: PropTypes.array,
  calname: PropTypes.string
};
