// @jacobmischka/ical-merger/dist/components/Subscription.js
function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

import _JSXStyle from "styled-jsx/style";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Clipboard from 'clipboard';
import Color from 'color';
import { COLORS, OPACITIES } from '../constants.js';
var buttonHoverBackgroundColor = new Color(COLORS.PRIMARY).alpha(OPACITIES.SECONDARY);
var secondaryText = new Color(COLORS.TEXT).alpha(OPACITIES.TEXT.SECONDARY);

var Subscription = /*#__PURE__*/function (_Component) {
  _inherits(Subscription, _Component);

  var _super = _createSuper(Subscription);

  function Subscription(props) {
    var _this;

    _classCallCheck(this, Subscription);

    _this = _super.call(this, props);
    _this.state = {
      showSub: false,
      copyButtonHovered: false,
      showUrl: false,
      copied: false
    };
    _this.handleShowSub = _this.handleShowSub.bind(_assertThisInitialized(_this));
    _this.handleHideSub = _this.handleHideSub.bind(_assertThisInitialized(_this));
    _this.toggleShowUrl = _this.toggleShowUrl.bind(_assertThisInitialized(_this));
    _this.handleCopyButtonMouseDown = _this.handleCopyButtonMouseDown.bind(_assertThisInitialized(_this));
    _this.handleCopyButtonMouseUp = _this.handleCopyButtonMouseUp.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(Subscription, [{
    key: "UNSAFE_componentWillUpdate",
    value: function UNSAFE_componentWillUpdate(nextProps, nextState) {
      if (this.state.showSub && !nextState.showSub || this.state.showUrl && !nextState.showUrl) this.clipboard.destroy();
    }
  }, {
    key: "render",
    value: function render() {
      var url = this.props.url;
      var dirs = url.split('?')[0].split('/');
      var filename = dirs[dirs.length - 1];
      var webcalUrl = "webcal://".concat(url.substring(url.indexOf('://') + 3));
      return /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "subscription-component"
      }, this.state.showSub ? /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]])
      }, /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "sub-controls"
      }, /*#__PURE__*/React.createElement("a", {
        id: "webcal-button",
        href: webcalUrl,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "button outline"
      }, "Subscribe with Outlook"), this.state.showUrl ? /*#__PURE__*/React.createElement("section", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "copy-with-url"
      }, /*#__PURE__*/React.createElement("input", {
        type: "text",
        id: "sub-url",
        value: url,
        readOnly: true,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]])
      }), /*#__PURE__*/React.createElement("button", {
        id: "copy-button",
        "data-clipboard-target": "#sub-url",
        title: "Long press to hide url",
        onMouseDown: this.handleCopyButtonMouseDown,
        onMouseUp: this.handleCopyButtonMouseUp,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "button"
      }, "Copy subscription address")) : /*#__PURE__*/React.createElement("button", {
        id: "copy-button",
        "data-clipboard-text": url,
        title: "Long press to show url",
        onMouseDown: this.handleCopyButtonMouseDown,
        onMouseUp: this.handleCopyButtonMouseUp,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "button outline"
      }, "Copy subscription address"), /*#__PURE__*/React.createElement("a", {
        id: "download-button",
        href: url,
        target: "_blank",
        rel: "noopener noreferrer",
        download: filename,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "button outline"
      }, "Download ICal/.ics file")), /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "text-center"
      }, /*#__PURE__*/React.createElement("a", {
        href: "#",
        id: "hide-sub",
        onClick: this.handleHideSub,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]])
      }, "Hide subscription info"))) : /*#__PURE__*/React.createElement("div", {
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "sub-controls"
      }, /*#__PURE__*/React.createElement("button", {
        onClick: this.handleShowSub,
        className: _JSXStyle.dynamic([["1389108452", [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]]]) + " " + "button outline"
      }, "Subscribe to this calendar")), /*#__PURE__*/React.createElement(_JSXStyle, {
        id: "1389108452",
        dynamic: [secondaryText, secondaryText, COLORS.PRIMARY, COLORS.PRIMARY, buttonHoverBackgroundColor, COLORS.PRIMARY, buttonHoverBackgroundColor]
      }, [".subscription-component.__jsx-style-dynamic-selector{margin:2em;}", ".sub-controls.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-wrap:wrap;-ms-flex-wrap:wrap;flex-wrap:wrap;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-box-pack:space-around;-webkit-justify-content:space-around;-ms-flex-pack:space-around;justify-content:space-around;-webkit-align-items:flex-end;-webkit-box-align:flex-end;-ms-flex-align:flex-end;align-items:flex-end;font-size:1.25em;}", ".copy-with-url.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;-webkit-align-items:stretch;-webkit-box-align:stretch;-ms-flex-align:stretch;align-items:stretch;}", "#sub-url.__jsx-style-dynamic-selector{display:block;-webkit-user-select:all;-moz-user-select:all;-ms-user-select:all;user-select:all;color:".concat(secondaryText, ";}"), "#hide-sub.__jsx-style-dynamic-selector{font-size:0.85em;-webkit-text-decoration:none;text-decoration:none;color:".concat(secondaryText, ";}"), "#hide-sub.__jsx-style-dynamic-selector:hover{-webkit-text-decoration:underline;text-decoration:underline;}", "input.__jsx-style-dynamic-selector{font-size:1.5em;padding:0.5em;border:1px solid rgba($grey-color,0.45);}", ".button.__jsx-style-dynamic-selector{margin:0.5em;border-radius:5px;padding:0.5em 1em;cursor:pointer;-webkit-text-decoration:none;text-decoration:none;color:white;font-size:1.25em;line-height:1.4;border:2px solid ".concat(COLORS.PRIMARY, ";background:").concat(COLORS.PRIMARY, ";}"), ".button.__jsx-style-dynamic-selector:hover{background:".concat(buttonHoverBackgroundColor, ";}"), ".button.outline.__jsx-style-dynamic-selector{background:transparent;color:".concat(COLORS.PRIMARY, ";}"), ".button.outline.__jsx-style-dynamic-selector:hover{color:white;background:".concat(buttonHoverBackgroundColor, ";}"), ".text-center.__jsx-style-dynamic-selector{text-align:center;}", "@media print{.subscription-component.__jsx-style-dynamic-selector{display:none;}}"]));
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      var _this2 = this;

      if (this.state.showSub) {
        this.clipboard = new Clipboard('#copy-button');
        this.clipboard.on('success', function () {
          _this2.setState({
            copied: true
          }, function () {
            window.setTimeout(function () {
              _this2.setState({
                copied: false
              });
            }, 2000);
          });
        });
      }
    }
  }, {
    key: "handleShowSub",
    value: function handleShowSub(event) {
      event.preventDefault();
      this.setState({
        showSub: true
      });
    }
  }, {
    key: "handleHideSub",
    value: function handleHideSub(event) {
      event.preventDefault();
      this.setState({
        showSub: false
      });
    }
  }, {
    key: "toggleShowUrl",
    value: function toggleShowUrl() {
      this.setState(function (state) {
        return {
          showUrl: !state.showUrl
        };
      });
    }
  }, {
    key: "handleCopyButtonMouseDown",
    value: function handleCopyButtonMouseDown() {
      var _this3 = this;

      this.copyButtonPressTimeout = window.setTimeout(function () {
        _this3.toggleShowUrl();
      }, 1000);
    }
  }, {
    key: "handleCopyButtonMouseUp",
    value: function handleCopyButtonMouseUp() {
      if (this.copyButtonPressTimeout) window.clearTimeout(this.copyButtonPressTimeout);
    }
  }]);

  return Subscription;
}(Component);

export { Subscription as default };
Subscription.propTypes = {
  url: PropTypes.string
};
